junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/TempDirectoryPerDeclarationTests.java

```diff
@@ -0,0 +1,981 @@
+	private static void assertSingleFailedContainer(EngineExecutionResults results,
+			Condition<Throwable>... conditions) {
+
+		results.containerEvents()//
+				.assertStatistics(stats -> stats.started(2).failed(1).succeeded(1))//
+				.assertThatEvents().haveExactly(1, finishedWithFailure(conditions));
+	}
+
+	private static void assertSingleFailedTest(EngineExecutionResults results, Class<? extends Throwable> clazz,
+			String message) {
+
+		assertSingleFailedTest(results, instanceOf(clazz), message(actual -> actual.contains(message)));
+	}
+
+	@SafeVarargs
+	@SuppressWarnings("varargs")
+	private static void assertSingleFailedTest(EngineExecutionResults results, Condition<Throwable>... conditions) {
+		results.testEvents().assertStatistics(stats -> stats.started(1).failed(1).succeeded(0));
+		results.testEvents().assertThatEvents().haveExactly(1, finishedWithFailure(conditions));
+	}
+
+	// -------------------------------------------------------------------------
+
+	static class AnnotationOnPrivateInstanceFieldTestCase {
+
+		@SuppressWarnings("unused")
+		@TempDir
+		private Path tempDir;
+
+		@Test
+		void test() {
+		}
+
+	}
+
+	static class AnnotationOnPrivateStaticFieldTestCase {
+
+		@SuppressWarnings("unused")
+		@TempDir
+		private static Path tempDir;
+
+		@Test
+		void test() {
+		}
+
+	}
+
+	static class AnnotationOnStaticFieldWithUnsupportedTypeTestCase {
+
+		@SuppressWarnings("unused")
+		@TempDir
+		static String tempDir;
+
+		@Test
+		void test1() {
+		}
+
+	}
+
+	static class AnnotationOnInstanceFieldWithUnsupportedTypeTestCase {
+
+		@SuppressWarnings("unused")
+		@TempDir
+		String tempDir;
+
+		@Test
+		void test1() {
+		}
+
+	}
+
+	static class AnnotationOnConstructorParameterTestCase {
+
+		AnnotationOnConstructorParameterTestCase(@SuppressWarnings("unused") @TempDir Path tempDir) {
+			// never called
+		}
+
+		@Test
+		void test() {
+			// never called
+		}
+
+	}
+
+	@TestInstance(PER_CLASS)
+	static class AnnotationOnConstructorParameterWithTestInstancePerClassTestCase
+			extends AnnotationOnConstructorParameterTestCase {
+
+		AnnotationOnConstructorParameterWithTestInstancePerClassTestCase(@TempDir Path tempDir) {
+			super(tempDir);
+		}
+	}
+
+	static class InvalidTestCase {
+
+		@Test
+		void wrongParameterType(@SuppressWarnings("unused") @TempDir String ignored) {
+			fail("this should never be called");
+		}
+	}
+
+	@Nested
+	@DisplayName("resolves java.io.File injection type")
+	class FileAndPathInjection {
+
+		@TempDir
+		File fileTempDir;
+
+		@TempDir
+		Path pathTempDir;
+
+		@Test
+		@DisplayName("and injected File and Path reference the same temp directory")
+		void checkFile(@TempDir File tempDir, @TempDir Path ref) {
+			assertNotNull(tempDir);
+			assertNotNull(ref);
+			assertNotNull(this.fileTempDir);
+			assertNotNull(this.pathTempDir);
+		}
+
+	}
+
+	// https://github.com/junit-team/junit5/issues/1748
+	static class TempDirectoryDoesNotPreventConstructorParameterResolutionTestCase {
```

Comment:
Isn't this one actually missing a `@TempDir` field or a `@ExtendWith(TempDirectory.class)` annotation?

Reply:
It worked anyway since the extension is registered globally but I think it's clearer if it actually used a `@TempDir` and added one in e79d45a65e60dfe6e4c089bdc26424d261942d58.